File: Graph/Traversal (DFS or BFS) Algo/Find Eventual Safe States/dfs_app.cpp
#include<bits/stdc++.h>
using namespace std;

//Time complexity :- O()
//Space complexity :- O()

// Approach :-
// All the nodes which are part of cycle or which leads to cycle will not be safe node 
// apart from that everyone can be safe nodes 
// Using dfs cycle detection technique we can mark vis & pathVis of the graph 
// which graph is not pathVis will be safe nodes 

//link :- 
// leetcode - https://leetcode.com/problems/find-eventual-safe-states/description/
// gfg - https://www.geeksforgeeks.org/problems/eventual-safe-states/1

class Solution {
  private:
    bool dfs(int node, vector<vector<int>> &graph, vector<bool> &vis, vector<bool> &pathVis){
        // mark the current node as visited and path visited 
        vis[node] = true;
        pathVis[node] = true;
        
        // traverse on the adj node of current node 
        for(auto adjNd: graph[node]){
            if(!vis[adjNd]){
                if(dfs(adjNd, graph, vis, pathVis))
                    return true;
            }
            else if(pathVis[adjNd])
                return true;
        }
        // backtrack pathVis as no loop found 
        pathVis[node] = false;
        return false;
    }
  
  public:
    vector<int> safeNodes(int V, vector<vector<int>>& edges) {
        // Code here
        vector<vector<int>> graph(V);
        
        // build graph
        for(auto edge: edges){
            int u = edge[0], v = edge[1];
            graph[u].push_back(v);
        }
        
        vector<bool> vis(V, false), pathVis(V, false);
        
        // cycle detection technique using dfs
        for(int i=0; i<V; ++i){
            if(!vis[i]){
                dfs(i, graph, vis, pathVis);
            }
        }
        
        vector<int> ans;
        // check which node where not part of loop
        for(int i=0; i<V; ++i){
            if(!pathVis[i]){
                ans.push_back(i);
            }
        }
        return ans;
    }
};

File: Graph/Traversal (DFS or BFS) Algo/Mother Vertex/dfs_app.java
#include<bits/stdc++.h>
using namespace std;

//Time complextiy :- O(V+E)
// Space complexity :- O(V)

// Approach :-  
// The vertex which in last visited can be a potential candidate for the mother vertex 
// if all the vertex are not visited in last then no mother vertex exists 


//Link :- https://www.geeksforgeeks.org/problems/mother-vertex/1

class Solution {
    private void dfs(int node, boolean[] visited, List<List<Integer>> graph){ 
        visited[node] = true;
        
        for(int adjNd: graph.get(node)){
            if(!visited[adjNd]){
                dfs(adjNd, visited, graph);
            }
        }
    }
    
    public int findMotherVertex(int V, int[][] edges) {
        List<List<Integer>> graph = new ArrayList<>();
        
        // initialize the adjacency list 
        for(int i=0; i<V; ++i){ 
            graph.add(new ArrayList<>());
        }
        
        // build the graph 
        for(int[] edge: edges){
            int u = edge[0];
            int v = edge[1];
            
            graph.get(u).add(v);
        }
        
        boolean[] visited = new boolean[V];
        int candidate = 0;
        
        // find the last finished vertex
        for(int i=0; i<V; ++i){
            if(!visited[i]){
                dfs(i, visited, graph);
                candidate = i;
            }
        }
        
        // check if the candidate reached all vertex
        Arrays.fill(visited, false);
        dfs(candidate, visited, graph);
    
        // validate if the candidate is mother vertex 
        for(boolean v : visited){
            if(!v)      return -1;
        }
        return candidate;
    }
}